### PR TITLE
[FEAT] Changed id type to Union{Int, String}

### PR DESCRIPTION
**Pull request by AzilAverakot**
Added functionality to handle String ids.

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -1,13 +1,14 @@
 name = "LightOSM"
 uuid = "d1922b25-af4e-4ba3-84af-fe9bea896051"
 authors = ["Jack Chan <jchan2@deloitte.com.au>"]
-version = "0.2.12"
+version = "0.2.13"
```
Comment (captchanjack) [inline Project.toml:4]:
```suggestion
version = "0.3.0"
```

File: src/constants.jl [hunk 2]
```diff
@@ -1,8 +1,8 @@
 """
 Default data types used to construct OSMGraph object.
 """
-const DEFAULT_OSM_ID_TYPE = Int64
-const DEFAULT_OSM_INDEX_TYPE = Int32
+const DEFAULT_OSM_INDEX_TYPE = Int64
```
Comment (captchanjack) [inline src/constants.jl:4]:
```suggestion
const DEFAULT_OSM_INDEX_TYPE = Int32
```

File: src/graph_utilities.jl [hunk 3]
```diff
@@ -25,6 +25,15 @@ Maps node id to index.
 node_id_to_index(g::OSMGraph, x::Integer) = g.node_to_index[x]
```
Comment (captchanjack) [inline src/graph_utilities.jl:25]:
```suggestion
node_id_to_index(g::OSMGraph, x::T <: Union{Int, String}) = g.node_to_index[x]
```

File: src/graph_utilities.jl [hunk 4]
```diff
@@ -25,6 +25,15 @@ Maps node id to index.
...
 node_id_to_index(g::OSMGraph, x::Vector{<:Integer}) = [node_id_to_index(g, i) for i in x]
 
+"""
+    node_id_to_index(g::OSMGraph, x::String)
+    node_id_to_index(g::OSMGraph, x::Vector{<:String})
+
+Maps node id to index.
+"""
+node_id_to_index(g::OSMGraph, x::String) = g.node_to_index[x]
+node_id_to_index(g::OSMGraph, x::Vector{<:String}) = [node_id_to_index(g, i) for i in x]
```
Comment (captchanjack) [inline src/graph_utilities.jl:35]: won't need this if `x` is typed parametrically

File: src/graph_utilities.jl [hunk 5]
```diff
@@ -40,27 +49,51 @@ node_to_index(g::OSMGraph, x::Vector{Node}) = [node_id_to_index(g, i.id) for i i
 Maps node index to dijkstra state (parents).
 """
 index_to_dijkstra_state(g::OSMGraph, x::Integer) = g.dijkstra_states[x]
+"""
+    index_to_dijkstra_state(g::OSMGraph, x::String)
+
+Maps node index to dijkstra state (parents).
+"""
+index_to_dijkstra_state(g::OSMGraph, x::String) = g.dijkstra_states[x]
```
Comment (captchanjack) [inline src/graph_utilities.jl:57]: same here

File: src/graph_utilities.jl [hunk 6]
```diff
@@ -40,27 +49,51 @@ node_to_index(g::OSMGraph, x::Vector{Node}) = [node_id_to_index(g, i.id) for i i
...
 
 """
     node_id_to_dijkstra_state(g::OSMGraph, x::Integer)
 
 Maps node id to dijkstra state (parents).
 """
 node_id_to_dijkstra_state(g::OSMGraph, x::Integer) = g.dijkstra_states[node_id_to_index(g, x)]
+"""
+    node_id_to_dijkstra_state(g::OSMGraph, x::String)
```
Comment (captchanjack) [inline src/graph_utilities.jl:66]: same here

File: src/graph_utilities.jl [hunk 7]
```diff
@@ -40,27 +49,51 @@ node_to_index(g::OSMGraph, x::Vector{Node}) = [node_id_to_index(g, i.id) for i i
...
 
 Set dijkstra state (parents) with node index.
 """
 set_dijkstra_state_with_index!(g::OSMGraph, index::Integer, state) = push!(g.dijkstra_states, index, state)
+"""
+    set_dijkstra_state_with_index!(g::OSMGraph, index::Integer, state)
+
+Set dijkstra state (parents) with node index.
+"""
+set_dijkstra_state_with_index!(g::OSMGraph, index::String, state) = push!(g.dijkstra_states, index, state)
```
Comment (captchanjack) [inline src/graph_utilities.jl:83]: same here

File: src/graph_utilities.jl [hunk 8]
```diff
@@ -40,27 +49,51 @@ node_to_index(g::OSMGraph, x::Vector{Node}) = [node_id_to_index(g, i.id) for i i
...
 
 Set dijkstra state (parents) with node id.
 """
 set_dijkstra_state_with_node_id!(g::OSMGraph, node_id::Integer, state) = push!(g.dijkstra_states, node_id_to_index(g, node_id), state)
+"""
+    set_dijkstra_state_with_node_id!(g::OSMGraph, index::String, state)
+
+Set dijkstra state (parents) with node id.
+"""
+set_dijkstra_state_with_node_id!(g::OSMGraph, node_id::String, state) = push!(g.dijkstra_states, node_id_to_index(g, node_id), state)
```
Comment (captchanjack) [inline src/graph_utilities.jl:96]: same here

File: src/graph_utilities.jl [hunk 9]
```diff
@@ -69,4 +102,6 @@ set_dijkstra_state_with_node_id!(g::OSMGraph, node_id::Integer, state) = push!(g
 Get maxspeed from index id or node id.
 """
 maxspeed_from_index(g, x::Integer) = index_to_node(g, x).tags["maxspeed"]
-maxspeed_from_node_id(g, x::Integer) = g.nodes[x].tags["maxspeed"]
\ No newline at end of file
+maxspeed_from_node_id(g, x::Integer) = g.nodes[x].tags["maxspeed"]
+maxspeed_from_index(g, x::String) = index_to_node(g, x).tags["maxspeed"]
+maxspeed_from_node_id(g, x::String) = g.nodes[x].tags["maxspeed"]
```
Comment (captchanjack) [inline src/graph_utilities.jl:107]: same here

File: src/nearest_node.jl [hunk 10]
```diff
@@ -48,13 +48,20 @@ not included in the results.
 """
 nearest_node(g::OSMGraph, node::Node) = nearest_node(g, node.location, (index)->index==g.node_to_index[node.id])
 nearest_node(g::OSMGraph, node_id::Integer) = nearest_node(g, g.nodes[node_id])
```
Comment (captchanjack) [inline src/nearest_node.jl:50]: same here


```suggestion
nearest_node(g::OSMGraph, node_id::T <: Union{..,..}) = nearest_node(g, node_to_index(g, node_id))
```

File: src/nearest_node.jl [hunk 11]
```diff
@@ -48,13 +48,20 @@ not included in the results.
...
+nearest_node(g::OSMGraph, node_id::String) = nearest_node(g, g.nodes[node_id])
 nearest_node(g::OSMGraph, nodes::Vector{<:Node}) = nearest_node(g, [n.id for n in nodes])
 function nearest_node(g::OSMGraph, node_ids::AbstractVector{<:Integer})
     locations = [g.nodes[n].location for n in node_ids]
     cartesian_locations = to_cartesian(locations)
     idxs, dists = knn(g.kdtree, cartesian_locations, 2, true)
     return [g.index_to_node[index[2]] for index in idxs], [d[2] for d in dists]
 end
+function nearest_node(g::OSMGraph, node_ids::AbstractVector{<:String})
```
Comment (captchanjack) [inline src/nearest_node.jl:59]: same here

File: src/nearest_way.jl [hunk 12]
```diff
@@ -123,3 +123,38 @@ function nearest_point_on_way(g::OSMGraph, point::GeoLocation, way_id::Integer)
...
+- `g::OSMGraph`: LightOSM graph.
+- `point::GeoLocation`: Point to find nearest position to.
+- `wid::Integer`: Way ID to search.
+
+# Returns
+- `::Tuple`:
+  - `::EdgePoint`: Nearest position along the way between two nodes.
+  - `::Float64`: Distance from `point` to the nearest position on the way.
+"""
+function nearest_point_on_way(g::OSMGraph, point::GeoLocation, way_id::String)
```
Comment (captchanjack) [inline src/nearest_way.jl:141]: same here

File: src/graph.jl [hunk 13]
```diff
@@ -52,12 +52,12 @@ function graph_from_object(osm_data_object::Union{XMLDocument,Dict};
     add_node_tags!(g)
     !(network_type in [:bike, :walk]) && add_indexed_restrictions!(g)
 
-    if precompute_dijkstra_states
-        add_dijkstra_states!(g)
-    else
-        U = DEFAULT_OSM_INDEX_TYPE
-        g.dijkstra_states = Vector{Vector{U}}(undef, length(g.nodes))
-    end
+    # if precompute_dijkstra_states
```
Comment (captchanjack) [inline src/graph.jl:55]: did you make a change here?

**Review by captchanjack — CHANGES_REQUESTED** [review]

File: src/types.jl [hunk 14]
```diff
@@ -40,12 +40,12 @@ end
 OpenStreetMap node.
 
 # Fields
-`T<:Integer`
+`T<:String`
```
[the remark this answers is not in the document]
Comment (AzilAverakot) [reply src/types.jl:43]: Can't set a type to a constant for types.jl since Julia doesn't allow this.

**Comment by captchanjack** [comment]
Lgtm

**Pull request merged**